src/Workspaces/Core/Portable/Workspace/Workspace_Events.cs

```diff
@@ -59,14 +77,25 @@ protected Task RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind kind, Solutio
             projectId = documentId.ProjectId;
         }
 
-        var ev = GetEventHandlers<WorkspaceChangeEventArgs>(WorkspaceChangeEventName);
+        var ev = GetEventHandlers<WorkspaceChangeEventArgs>(WorkspaceChangedImmediateEventName);
+        WorkspaceChangeEventArgs args = null;
+        if (ev.HasHandlers)
+        {
+            using (Logger.LogBlock(FunctionId.Workspace_EventsImmediate, (s, p, d, k) => $"{s.Id} - {p} - {d} {kind.ToString()}", newSolution, projectId, documentId, kind, CancellationToken.None))
+            {
+                args = new WorkspaceChangeEventArgs(kind, oldSolution, newSolution, projectId, documentId);
+                ev.RaiseEvent(static (handler, arg) => handler(arg.self, arg.args), (self: this, args));
+            }
+        }
```

Comment:
Can we extract this bit into a local function so there's not repetition?

Reply:
Extracted into local method, slightly complicated by wanting to keep the args allocation to be just once